lib/generate-critical-css.js

```diff
@@ -8,12 +10,15 @@ async function generateCriticalCSS( {
 	urls,
 	viewports,
 	filters,
+	meta,
 } ) {
 	if ( ! progressCallback ) {
 		progressCallback = () => {};
 	}
 
 	try {
+
+		urls = await getValidUrls(urls, meta, browserInterface);
```

Comment:
🤔  Now I see this implementation, I don't think we need a 3rd URL loading pass upfront to check for URL validity.

In this PR, you're passing in a minimum and maximum number of successful page-loads / generation attempts. It should be easy to loop through all provided URLs in the [main loop](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/trunk/lib/generate-critical-css.js#L41) and abort early once enough urls have been successfully processed. At the end of the loop, it can check if enough urls were successfully processed and throw an error if not.

Reply:
@thingalon ,

I have been struggling with it which is that the `loadPage` [method ](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/browser-interface-iframe.js#L120) is using throw statement to stop the whole process early if an url in the pool of urls has an error. 

Since `loadPage` is effectively called 3 times in the `generateCriticalCSS` [function](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L7) via 

A: `await CSSFileSet.collate( browserInterface, urls );` ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L26))
B: `const pageSelectors = await browserInterface.runInPage(` ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L48))
C: const pageAboveFold = await browserInterface.runInPage(' ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L76))

There are actually 2 loops too, the one in **A**, and the one that encompass **B** and **C**.

Regardless of the 2 loops which was one hurdle for me, the `throw` statements from the `loadPage` which is bubbled up to the `generateCriticalCSS` was the second hurle. If we have 10 urls and want 5 successful one, if 1,2,3,4 urls are failing, we still want to carry on the process for url 5, 6, 7, 8 and 9.

So I decided to use `getValidUrls` to go throw the whole pool of urls and identify the valid and invalid one without raising an error and just generating the valid set of urls which can then be used in the next 2 loops.

Not sure if it makes sense, I might still not be seeing the easier path for implementation here without a bigger refactoring. I think at this stage it would be useful for me to have a video chat and session sharing with you.